h2/src/main/org/h2/engine/Session.java

```diff
@@ -312,7 +308,7 @@ public Value getVariable(String name) {
         if (variables == null) {
             return new String[0];
         }
-        return variables.keySet().toArray(new String[variables.size()]);
+        return variables.keySet().toArray(new String[0]);
```

Comment:
Old code was better, it's not the `ArrayList` that works faster with an empty array.

Reply:
I was simply following IDEA's suggestion to make code "green", and I do not remember any distinctions between ArrayList and other popular collections made  [here](https://shipilev.net/blog/2016/arrays-wisdom-ancients/#_conclusion).

Reply:
I tested different collections with JMH on different Java versions about a year ago. I don't remember the exact results, but other collections were typically slower.

IDEA can suggest bad replacements sometimes (and other IDEs too).

I can retest it to be sure.

Reply:
On Java 13 `HashMap.keySet().toArray(T[])` shows surprising results:
```
       (size)         Score         Error  Units
sized       0  71712214,700 ± 1243641,226  ops/s
sized       1  12856321,292 ± 2287814,065  ops/s
sized      10   5074629,159 ±  522772,443  ops/s
sized     100    566289,897 ±   16597,121  ops/s
sized    1000     39620,832 ±     887,907  ops/s
zero        0  98146908,786 ±  532451,125  ops/s
zero        1  13203083,218 ±  309671,958  ops/s
zero       10   4778969,230 ± 1488026,965  ops/s
zero      100    519223,337 ±   14512,372  ops/s
zero     1000     49107,862 ±     788,980  ops/s
```
Empty array is faster for 0, 1, and 1000 elements and proper array is faster for 10 and 100 elements.

Reply:
For `HashSet` zero size should be good enough, it is only slightly slower with few elements, and in other cases (zero or many) it is slightly faster.

Reply:
And I could not agree more with the last section of [this article](https://shipilev.net/blog/2016/arrays-wisdom-ancients/#_parting_thoughts). 
>most of the time, the straight-forward code is fast enough, so stop philosophizing about angels on the head of a pin, and get back to work

:smile: